nautobot/docs/development/best-practices.md

```diff
@@ -63,3 +63,193 @@ class ExampleModel(PrimaryModel):
+  # falls back to the default of "exact".
+  def _console_ports(self, queryset, name, value):
+      breakpoint()  # This is where we'll illustrate pdb below
+      return queryset.exclude(consoleports__isnull=value)
+```
+
+The default `lookup_expr` unless otherwise specified is “exact”, as seen in [django_filters.conf](https://github.com/carltongibson/django-filter/blob/main/django_filters/conf.py#L10):
+
+```python
+  'DEFAULT_LOOKUP_EXPR': 'exact',
+```
+
+When this method is called, the internal state is default, making reverse introspection impossible, because the `lookup_expr` is defaulting to “exact”:
+
+```python
+(Pdb) field = self.filters["console_ports"]
+(Pdb) field.exclude
+False
+(Pdb) field.lookup_expr
+'exact' 
+```
+
+This means that the arguments for the field are being completely ignored and the hard-coded queryset `queryset.exclude(consoleports__isnull=value)` is all that is being run when this method is called. This hard-coding is impossible to introspect and therefore reverse.
+
+In fact, this method is completely unnecessary, because it could be replaced with options on the filter field itself:
+
+```python
+    console_ports = django_filters.BooleanFilter(
+        field_name="consoleports",  # The actual related field name
+        exclude=True,               # Perform an `.exclude()` vs. `.filter()``
+        lookup_expr="isnull",       # Perform `isnull` vs. `exact``
+        label="Has console ports",
+    )
+```
+
+Now, if we use another breakpoint (this time inside of the `DeviceListView.get()` since there’s no method we can break into), you can see that the filter field now has the correect attributes that can be used to accurately reverse this query:
+
+```python
+(Pdb) filterset = self.filterset(request.GET, self.queryset)
+(Pdb) field = filterset.filters["console_ports"]
+(Pdb) field.exclude
+True
+(Pdb) field.lookup_expr
+'isnull'
+```
+
+#### Generating Reversible Q Objects
+
+These field values could be used to construct a `Q()` query that looks something like:
+
+```python
+def generate_query(self, field_name, lookup_expr, value):
+    query = Q()
+    predicate = {f"{field_name}__{lookup_expr}": value}
+    if field.exclude:
+        query |= ~Q(**predicate)
+    else:
+        query |= Q(**predicate)
+    return query
+    
+## Somewhere else in business logic:
+field = filterset.filters[name]
+value = filterset.data[name]
+query = generate_query(field.field_name, field.lookup_expr, value)
+filterset.qs.filter(query).count()  # 339
+```
+
+But furthermore, this also becomes unnecessary for the common case, because we can **just trust the filterset itself to do the proper filtering and completely obviate the need to reverse the query at all.**
+
+Consider this input filter:
+
+```python
+filter_params = {"console_ports": True}
+```
+
+With properly configured filter fields, this will just work:
+
+```python
+DeviceFilterSet(filter_params).qs.count()  # 339
+```
```

Comment:
Not sure I understand what this section is arguing for or against, or guiding me to do or not do as far as coding best practices. It seems to be saying "Hey, you could do this, but it would be redundant" - so what's the takeaway?

Reply:
This entire section of the document could probably just be left out or replaced since it was a part of my initial analysis/discovery. 

On that note, it almost feels like my [response to comment in above](https://github.com/nautobot/nautobot/pull/1691#discussion_r863167001) could be what replaces it. Thoughts there?

The other thing that comes to mind is more of a "what if" (similarly to the slug/id filter field) we implemented a "composite" field that took arguments of the various field names (such as in the case of `DeviceFilterSet.pass_through_ports` that could facilitate such composite lookups but still provide arguments/attributes that could be use to inform query construction.